filesmanager/filesmanager.py

```diff
@@ -570,15 +590,38 @@ def upload_file_to_directory(self, file, target_directory):
             from cms.djangoapps.contentstore.asset_storage_handler import \
                 update_course_run_asset  # pylint: disable=import-outside-toplevel
 
-        file_object = self.temporary_uploaded_files.pop(file.get("name"), None)
+        file_object = self.find_temporary_file(file.get("name"))
         file_path, name = file.get("path"), file.get("name")
         metadata = file.get("metadata", {})
 
+        if "Unpublished" in file_path:
+            print("File moved from other Folder to Unpublished", metadata)
+            if metadata.get("from"):
+                del self.source_keys[metadata.get("from")]
+            return
         if file_object:
             file_path, name = self.generate_content_path(file_path, file_object.filename)
-            file_object.file._set_name(name) # pylint: disable=protected-access
+            file_object.file._set_name(self.generate_asset_name(file_path)) # pylint: disable=protected-access
             content = update_course_run_asset(self.course_id, file_object.file)
             metadata = self.get_asset_json_from_content(content)
+        if not metadata.get("uploaded_at"):
+            file_path, name = self.generate_content_path(file_path, metadata.get("display_name"))
+            source_asset_key = metadata.get("asset_key")
+            location = AssetKey.from_string(source_asset_key)
+            content = contentstore().find(location)
+            tmp_file = tempfile.NamedTemporaryFile(suffix=content.content_type.split("/")[-1])
+            tmp_file.write(content._data)
+            tmp_file.file.seek(0)
+            memory_file = InMemoryUploadedFile(file=tmp_file, field_name=None, name=metadata.get("display_name"), content_type=content.content_type, size=content.length, charset=None)
+
+            memory_file._set_name(self.generate_asset_name(file_path)) # pylint: disable=protected-access
+            content = update_course_run_asset(self.course_id, memory_file)
+            metadata = self.get_asset_json_from_content(content)
+            metadata["from"] = source_asset_key
+
+            self.source_keys[source_asset_key] = metadata.get("asset_key")
```

Comment:
This is pretty hard to understand, can we leave an inline comment explaining why we are doing this? can we move it to a different method?